### ISSUE TITLE
LightGBMTuner vs "regular Optuna"

### ISSUE THREAD

**Issue by michael135** [optuna/optuna#2102]
I follow this example,
https://medium.com/optuna/lightgbm-tuner-new-optuna-integration-for-hyperparameter-optimization-8b7095e99258

If I understand correctly in above example LightGBMTuner is not used ?
Is there any example example of LightGBMTuner usage?

What is an advantage of using LightGBMTuner instead of "regular Optuna"?

At (https://optuna.readthedocs.io/en/stable/faq.html#how-to-save-machine-learning-models-trained-in-objective-functions) it's nice example! But if I do 10K trials, somehow I'm not excited to save 10K models, which can take a few Gb.

**Comment by crcrpar** [comment]
> If I understand correctly in above example LightGBMTuner is not used ?
> Is there any example example of LightGBMTuner usage?

LightGBMTuner is aimed to be a drop-in replacement for LightGBM. It uses Optuna inside. The implementation is not so simple but you can check this by searching "study.optimize" in https://github.com/optuna/optuna/blob/master/optuna/integration/_lightgbm_tuner/optimize.py.

> What is an advantage of using LightGBMTuner instead of "regular Optuna"?

One advantage is that the algorithm of LightGBMTuner is based on the kind of rule of thumb which is widely known in Kaggle community.

**Comment by michael135** [comment]
Thank you @crcrpar!

Can you elaborate in a few words about a "rule of thumb"?

**Comment by crcrpar** [comment]
I'm sorry, but it's beyond my description, and it's elaborated in **Step-wise algorithm** section of the medium post you mentioned.

**Comment by michael135** [comment]
@crcrpar Great. I will take a double read.

But at medium posts code, it doesn't use a `LightGMBTuner`.
Am I wrong?

**Comment by crcrpar** [comment]
Ah, I now see your question eventually I think. Sorry for bothering you. When we want to use `LightGBMTuner`, we basically don't use it directly in our code. We basically use it by replacing `import lightgbm as lgbm` in our code with `import optuna.integration.lightgbm as lgbm` (see the capture below) and then just call `lgbm.train` like the usual lightgbm code. With this replacement, `lgbm.train` makes use of `LightGBMTuner` internally and optimizes the hyperparameters.

Does this make sense to you? 🙂 

![image](https://user-images.githubusercontent.com/16191443/102063341-41a01400-3e39-11eb-8d75-952c30426a8a.png)

**Comment by crcrpar** [comment]
> At (https://optuna.readthedocs.io/en/stable/faq.html#how-to-save-machine-learning-models-trained-in-objective-functions) it's nice example! But if I do 10K trials, somehow I'm not excited to save 10K models, which can take a few Gb.

One way to select which model to keep or not is using callbacks for `study.optimize` (https://optuna.readthedocs.io/en/stable/reference/generated/optuna.study.Study.html#optuna.study.Study.optimize). See https://colab.research.google.com/drive/1IrJaoS7Ba85WYHSXXkhqFwp8EeBLUvDw?usp=sharing for the details. 😃

**Comment by michael135** [comment]
Absolutely 🙂.
Now I got it.

Thank you @crcrpar
